accounting/invoice/utils.py

```diff
@@ -21,27 +21,102 @@
 Invoice utilities.
 """
 
-from datetime import timedelta
+import datetime
 
-from django.utils import timezone
+from django.conf import settings
+from django.core.mail.message import EmailMessage
+from django.template.loader import get_template
 
+from accounting.account.models import Account
+from accounting.invoice import choices, models
 
-def get_last_day_past_month():
+
+def resolve_dict_callables(dictionary, *args):
     """
-    Return the `datetime` object representing the last day of the month that just passed.
+    Resolve any potential callables in a dictionary by passing it some arguments.
     """
-    return timezone.now().replace(day=1) - timedelta(days=1)
+    # We use a new dictionary to prevent mutating the callable. If we didn't, then on another round of calling this
+    # with the same dictionary, the callable would have disappeared.
+    new_dict = {}
+    for key, potential_callable in dictionary.items():
+        if callable(potential_callable):
+            new_dict[key] = potential_callable(*args)
+        else:
+            new_dict[key] = dictionary[key]
+    return new_dict
 
 
-def get_first_day_past_month():
-    """
-    Return the `datetime` object representing the first day of the month that just passed.
+# pylint: disable=too-many-arguments,too-many-locals
+def send_email_with_invoice(template, subject, extra_email_context=None, client_usernames=settings.BILLING_CYCLE_USERS,
+                            create_invoice=False, draft_invoice=False, fill_line_items_from_jira=False,
+                            upload_to_google_drive=False, auto_approve=False):
     """
-    return get_last_day_past_month().replace(day=1)
+    Send an email with an invoice attached, and potentially perform peripheral actions.
 
+    TODO: This function's pretty beefy; break it up by encapsulating logic into more utility functions or a new class.
```

Comment:
@UmanShahzad Are you planning on doing this in the context of the current task (https://tasks.opencraft.com/browse/OC-3669)? If none of the tasks that are scheduled for the next couple sprints are going to involve touching this code, it would probably be best to do this refactoring now.

Reply:
It's likely that I'll be touching this code for https://tasks.opencraft.com/browse/OC-3697, so I put this TODO here for that.